### DISCUSSION TITLE
Multiple power descriptors

### DISCUSSION THREAD

**Discussion by MikeLoh** [hathach/tinyusb#1170] [Q&A]
Has anyone had success in creating descriptors with multiple power settings, like 500mA and 0mA, letting the host to select. Then, is there a callback or other method to determine which descriptor was selected?

Want to have ability for the device to be connected to either a desktop or mobile without issue and operate on battery power when on mobile.

**Comment by aspdigital** [comment]
This is not possible and is not in the USB spec. How could the host decide what current draw a device should be allowed? Remember the device tells the host what current it needs, and then the host decides whether the port can source enough current to satisfy the device.

The only sort of work-around for this is would be for, at power-up, your device has to determine what current it requires from the bus. Maybe it can detect a plugged-in power supply? Once it knows, it can fill in the Configuration Descriptor's bMaxPower field and its bmAttrib field with the "self-power" bit set or cleared. Then it can allow enumeration to go forward. 

The real question is are you shutting off power to individual bits of the design if you need to keep power consumption to a minimum? If so, then again you need to do that before enumeration so you can adjust the descriptor. You can't change this after enumeration.

**Comment by aspdigital** [comment]
But I must make clear: the host does not decide how much current the device can draw on its own. It can't. There is nothing in the spec that allows for such choice.

The host is told how much current the device draws, and as part of enumeration, it checks to see how much power other devices on the bus are drawing, and if the new device pulls too much, the host will not allow that new device to enumerate.

**Comment by MikeLoh** [comment]
It has been a while since I was deep into USB spec, I thought this is possible. I guess then it would be a procedure:

Try at 500 mA
if not enumerated and connection fails
Try at 100 mA
if not enumerated and connection fails
Try at 0 mA (or whatever lowest, 8 mA for embedded host?)

Thanks for feedback.

**Reply by hathach** [reply]
unfortunately host driver does not work that way, it will either 
- skip enumerate device with high power in descriptor and/or
- enumerate but does not trust device at all in the power declaration and keep monitoring the current draw and will forcefully power off the port when device cross its threshold (iOS device is actively monitor this)

IMHO, the best approach is presenting **all the possible configurations to host with  multiple configurations descriptors** similar to this net example which https://github.com/hathach/tinyusb/blob/master/examples/device/net_lwip_webserver/src/usb_descriptors.c. Which present both rndis and ecm. Windows will pick rndis, macOS pick ecm, and Linux can pick either. Host based on its own logic will decide which to mount.

It also make sense that each (power) configuration has different interfaces, e.g skip msc and use hid only in lower configuration etc to keep the promise with the host.

**Reply by aspdigital** [reply]
@hathach is correct, @MikeLoh -- the host doesn't "step down" and try other power levels. 

Again, you _must_ understand this: as part of enumeration, your device tells the host how much current it draws.  The host keeps track of the current draw of all devices on the bus, and if you say "I need 500 mA" and the port has only 400 mA available, then the host will not allow your device to attach to the bus.  I know that macOS will do exactly that.

The host _will not_ go back and ask if there is an alternative lower-power configuration. Your best bet is to do as I suggested above, which is to have some way of determining prior to enumeration which parts of the circuit you wish to have operational and set the power field in the device descriptor appropriately.

And don't forget: if your goal really is to reduce power consumption by shutting off parts of your circuit, you must actually do that.

**Reply by MikeLoh** [reply]
Thank you for feedback and links. The goal is best operation with devices like iOS with camera connection kit or other handheld configurations. It is a device that has battery so when plugged into a Mac / PC we will want to charge, but when plugged into iOS we want to avoid the not supported message and operate on battery.

Thank you both for useful information. We will consider our options.

**Reply by amowry** [reply]
Did you ever find a good solution to this? I'm trying to do exactly the same thing, i.e. only charge the battery on my device if I'm connected to a host that allows enough current. I'd be curious to hear what the options are.
